File: 03.Whited/InfluenceFunctions.jl
"""
    InfluenceFunctions

Conditional moments, skewness, and their influence functions.
Port of influencefunctions.R. All influence functions return an n×1 matrix,
except `beta_infl`, which returns n×k. Estimators return scalars, except
`beta`, which returns a k×1 matrix.
"""
module InfluenceFunctions

using LinearAlgebra
using Statistics

export mean_cond, cov_cond, beta, skew, skew_cond,
       mean_infl, beta_infl, cov_infl, mean_cond_infl,
       mean_ratio_infl, cov_cond_infl, skew_infl, skew_cond_infl

_vec(a::AbstractVector) = float.(a)
_vec(a::AbstractMatrix) = float.(vec(a))
_col(a) = reshape(_vec(a), :, 1)
_mat(a::AbstractVector) = reshape(float.(a), :, 1)
_mat(a::AbstractMatrix) = float.(a)

# Conditional mean and covariance ---------------------------------------------

"""E(x | dum = 1), dum ∈ {0, 1}."""
function mean_cond(x, dum)
    x, dum = _vec(x), _vec(dum)
    return mean(x .* dum) / mean(dum)
end

"""cov(x, y | dum = 1), dum ∈ {0, 1}."""
function cov_cond(x, y, dum)
    x, y, dum = _vec(x), _vec(y), _vec(dum)
    μ_dum = mean(dum)
    return mean(x .* y .* dum) / μ_dum -
           (mean(x .* dum) / μ_dum) * (mean(y .* dum) / μ_dum)
end

"""Slope of the linear regression y = x * β + u."""
function beta(x, y)
    X, yv = _mat(x), _col(y)
    return (X' * X) \ (X' * yv)
end

"""E[(x - E[x])^3] / E[(x - E[x])^2]^(3/2)."""
function skew(x)
    x = _vec(x)
    d = x .- mean(x)
    return mean(d .^ 3) / mean(d .^ 2)^1.5
end

"""Conditional skewness, dum ∈ {0, 1}."""
function skew_cond(x, dum)
    x, dum = _vec(x), _vec(dum)
    μ_dum = mean(dum)
    μc_x3 = mean(x .^ 3 .* dum) / μ_dum
    μc_x2 = mean(x .^ 2 .* dum) / μ_dum
    μc_x = mean(x .* dum) / μ_dum
    F_dum = μc_x3 - 3 * μc_x2 * μc_x + 2 * μc_x^3
    G_dum = μc_x2 - μc_x^2
    return F_dum / G_dum^1.5
end

# Influence functions ---------------------------------------------------------

"""Influence function for the mean."""
function mean_infl(x)
    x = _vec(x)
    return _col(x .- mean(x))
end

"""Influence function for OLS coefficients in y = x * β + u."""
function beta_infl(x, y)
    X, yv = _mat(x), _vec(y)
    n = length(yv)
    b = (X' * X) \ (X' * yv)
    u_hat = yv .- X * b
    # infl = (X .* u_hat) * inv(X'X / n)
    return (X .* u_hat) / ((X' * X) ./ n)
end

"""Influence function for cov(x, y)."""
function cov_infl(x, y)
    x, y = _vec(x), _vec(y)
    μ_xy, μ_x, μ_y = mean(x .* y), mean(x), mean(y)
    return _col(x .* y .- μ_xy .- μ_y .* (x .- μ_x) .- μ_x .* (y .- μ_y))
end

"""Influence function for E(x | dum = 1), dum ∈ {0, 1}."""
function mean_cond_infl(x, dum)
    x, dum = _vec(x), _vec(dum)
    μ_dum = mean(dum)
    μ_x_dum = mean(x .* dum)
    return _col((μ_dum .* (x .* dum .- μ_x_dum) .-
                 μ_x_dum .* (dum .- μ_dum)) ./ μ_dum^2)
end

"""Influence function for E(x) / E(y)."""
function mean_ratio_infl(x, y)
    x, y = _vec(x), _vec(y)
    μ_x, μ_y = mean(x), mean(y)
    return _col((μ_y .* (x .- μ_x) .- μ_x .* (y .- μ_y)) ./ μ_y^2)
end

"""Influence function for cov(x, y | dum = 1), dum ∈ {0, 1}."""
function cov_cond_infl(x, y, dum)
    x, y, dum = _vec(x), _vec(y), _vec(dum)
    μ_dum = mean(dum)
    μ_x_y_dum = mean(x .* y .* dum)
    μ_x_dum = mean(x .* dum)
    μ_y_dum = mean(y .* dum)
    infl = (μ_dum .* (x .* y .* dum .- μ_x_y_dum) .-
            μ_x_y_dum .* (dum .- μ_dum)) ./ μ_dum^2 .-
           μ_y_dum / μ_dum .* (μ_dum .* (x .* dum .- μ_x_dum) .-
            μ_x_dum .* (dum .- μ_dum)) ./ μ_dum^2 .-
           μ_x_dum / μ_dum .* (μ_dum .* (y .* dum .- μ_y_dum) .-
            μ_y_dum .* (dum .- μ_dum)) ./ μ_dum^2
    return _col(infl)
end

"""Influence function for skewness."""
function skew_infl(x)
    x = _vec(x)

    μ_x3 = mean(x .^ 3)
    μ_x2 = mean(x .^ 2)
    μ_x = mean(x)
    F = μ_x3 - 3 * μ_x2 * μ_x + 2 * μ_x^3
    G = μ_x2 - μ_x^2

    ψ_x = x .- μ_x
    ψ_x2 = x .^ 2 .- μ_x2
    ψ_x3 = x .^ 3 .- μ_x3
    ψ_F = ψ_x3 .- 3 .* (ψ_x2 .* μ_x .+ μ_x2 .* ψ_x) .+ 6 * μ_x^2 .* ψ_x
    ψ_G = ψ_x2 .- 2 * μ_x .* ψ_x

    return _col((ψ_F .* G^1.5 .- 1.5 * F * G^0.5 .* ψ_G) ./ G^3)
end

"""Influence function for conditional skewness, dum ∈ {0, 1}."""
function skew_cond_infl(x, dum)
    x, dum = _vec(x), _vec(dum)

    x_dum = x .* dum
    x2_dum = x .^ 2 .* dum
    x3_dum = x .^ 3 .* dum

    μ_dum = mean(dum)
    μ_x3_dum = mean(x3_dum)
    μ_x2_dum = mean(x2_dum)
    μ_x_dum = mean(x_dum)
    μc_x3 = μ_x3_dum / μ_dum
    μc_x2 = μ_x2_dum / μ_dum
    μc_x = μ_x_dum / μ_dum
    F_dum = μc_x3 - 3 * μc_x2 * μc_x + 2 * μc_x^3
    G_dum = μc_x2 - μc_x^2

    ψ_x_dum = ((x_dum .- μ_x_dum) .* μ_dum .-
               μ_x_dum .* (dum .- μ_dum)) ./ μ_dum^2
    ψ_x2_dum = ((x2_dum .- μ_x2_dum) .* μ_dum .-
                μ_x2_dum .* (dum .- μ_dum)) ./ μ_dum^2
    ψ_x3_dum = ((x3_dum .- μ_x3_dum) .* μ_dum .-
                μ_x3_dum .* (dum .- μ_dum)) ./ μ_dum^2
    ψ_F_dum = ψ_x3_dum .-
              3 .* (ψ_x2_dum .* μ_x_dum .+ μ_x2_dum .* ψ_x_dum) .+
              6 * μ_x_dum^2 .* ψ_x_dum
    ψ_G_dum = ψ_x2_dum .- 2 * μ_x_dum .* ψ_x_dum

    return _col((ψ_F_dum .* G_dum^1.5 .-
                 1.5 * F_dum * G_dum^0.5 .* ψ_G_dum) ./ G_dum^3)
end

end # module

File: 03.Whited/influencefunctions.R
# Define conditional mean and covariance ----
mean_cond <- function(x, dum) {
    # Conditional mean    
    # E(x | dum = 1), dum = {0, 1}
    res <- mean(x*dum)/mean(dum)
    return(res)
}
cov_cond <- function(x, y, dum) {
    # Conditional covariance
    # cov(x, y | dum = 1), dum = {0, 1}
    res <- mean(x*y*dum)/mean(dum) - (mean(x*dum)/mean(dum)) * (mean(y*dum)/mean(dum))
    return(res)
}     
beta <- function(x, y) {
    # Returns slope in a linear regression
    # y = x \beta + u
    res <- solve(t(x) %*% x) %*% t(x) %*% y
    return(res)
}
skew <- function(x) {
    # Skewness
    # E[(x - E[x])^3]/E[(x - E[x])^2]^3/2
    res <- mean((x - mean(x))^3)/(mean((x - mean(x))^2))^(3/2)
    return(res)
}
skew_cond <- function(x, dum) {
    # Conditional skewness (see note_on_skewness)
    # E[(x - E[x])^3 | dum = 1]/E[(x - E[x])^2 | dum = 1]^3/2, dum = {0, 1}
    mu_cond_x3_dum <- mean(x^3*dum)/mean(dum)
    mu_cond_x2_dum <- mean(x^2*dum)/mean(dum)
    mu_cond_x_dum <- mean(x*dum)/mean(dum)
    F_dum <- mu_cond_x3_dum - 3*mu_cond_x2_dum*mu_cond_x_dum + 2*mu_cond_x_dum^3
    G_dum <- mu_cond_x2_dum - mu_cond_x_dum^2
    res <- F_dum/(G_dum**(3/2))
    return(res)
}
# Define influence functions ----
mean_infl <- function(x) {
    # Influence function for mean
    infl <- as.matrix(x - mean(x), ncol = 1) 
    return(infl)
}
beta_infl <- function(x, y) {
    # Influence function for OLS coefficients
    # y = x \beta + u
    
    # Convert to numeric matrices
    x <- data.matrix(x)
    y <- data.matrix(y)
    
    # Compute influence functions for beta
    n <- length(y)
    beta <- solve(t(x) %*% x) %*% t(x) %*% y
    u_hat <- y - x %*% beta
    u_hat_mtx <- matrix(u_hat, nrow = length(u_hat), ncol=ncol(x), byrow = FALSE)
    infl <- t(solve(t(x) %*% x/n) %*% t(x*u_hat_mtx))
    return(infl)
}
cov_infl <- function(x, y) {
    # Influence function for covariance
    # cov(x, y)
    mu_xy <- mean(x*y)
    mu_x <- mean(x)
    mu_y <- mean(y)
    infl <- x*y - mu_xy - mu_y*(x - mu_x) - mu_x*(y - mu_y)
    infl <- as.matrix(infl, ncol = 1) 
    return(infl)
}
mean_cond_infl <- function(x, dum) {
    # Influence function for the conditional mean
    # E(x | dum = 1) with dum = {0, 1}
    mu_dum <-  mean(dum)
    mu_x_dum <-  mean(x*dum)
    infl <- (mu_dum*(x*dum - mu_x_dum) - mu_x_dum*(dum - mu_dum))/(mu_dum^2)
    infl <- as.matrix(infl, ncol = 1) 
    return(infl)
}
mean_ratio_infl <- function(x, y) {
    # Influence function for the ratio of means
    # E(x)/E(y)
    mu_x <-  mean(x)
    mu_y <-  mean(y)
    infl <- (mu_y*(x - mu_x) - mu_x*(y - mu_y))/(mu_y^2)
    infl <- as.matrix(infl, ncol = 1) 
    return(infl)
}
cov_cond_infl <- function(x, y, dum) {
    # Influence function for the conditional covariance
    # cov(x, y | dum = 1) with dum = {0, 1}
    mu_dum <-  mean(dum)
    mu_x_y_dum <- mean(x*y*dum)
    mu_x_dum <- mean(x*dum)
    mu_y_dum <- mean(y*dum)
    infl <- ((mu_dum*(x*y*dum - mu_x_y_dum) - mu_x_y_dum*(dum - mu_dum))/(mu_dum^2)
             - mu_y_dum/mu_dum*(mu_dum*(x*dum - mu_x_dum) - mu_x_dum*(dum - mu_dum))/(mu_dum^2)
             - mu_x_dum/mu_dum*(mu_dum*(y*dum - mu_y_dum) - mu_y_dum*(dum - mu_dum))/(mu_dum^2))
    infl <- as.matrix(infl, ncol = 1) 
    return(infl)
}
skew_infl <- function(x) {
    # Influence function on skewness (see note_on_skewness)
    # E[(x - E[x])^3]/E[(x - E[x])^2]^3/2
    
    # Compute moments
    mu_x3 <- mean(x^3)
    mu_x2 <- mean(x^2)
    mu_x  <- mean(x)
    F <- mu_x3 - 3*mu_x2*mu_x + 2*mu_x^3
    G <- mu_x2 - mu_x^2
    
    # Compute basic influence functions
    psi_x <- x - mean(x)
    psi_x2 <- x^2 - mean(x^2)
    psi_x3 <- x^3 - mean(x^3)
    psi_F <- psi_x3 - 3*(psi_x2*mu_x + mu_x2*psi_x) + 6*mu_x^2*psi_x
    psi_G <- psi_x2 - 2*mu_x*psi_x
    
    infl <- (psi_F*G^(3/2) - 3/2*F*G^(1/2)*psi_G)/(G^3)
    infl <- as.matrix(infl, ncol = 1) 
    return(infl)
}
skew_cond_infl <- function(x, dum) {
    # Influence function on conditional skewness (see note_on_skewness)
    # E[(x - E[x])^3|D]/E[(x - E[x])^2|D]^3/2
    
    x_dum <- x*dum
    x2_dum <- x^2*dum
    x3_dum <- x^3*dum
    
    # Compute moments
    mu_dum <- mean(dum)
    mu_x3_dum <- mean(x3_dum)
    mu_x2_dum <- mean(x2_dum)
    mu_x_dum <- mean(x_dum)
    mu_cond_x3_dum <- mean(x^3*dum)/mean(dum)
    mu_cond_x2_dum <- mean(x^2*dum)/mean(dum)
    mu_cond_x_dum <- mean(x*dum)/mean(dum)
    F_dum <- mu_cond_x3_dum - 3*mu_cond_x2_dum*mu_cond_x_dum + 2*mu_cond_x_dum^3
    G_dum <- mu_cond_x2_dum - mu_cond_x_dum^2
    
    # Compute basic influence functions
    psi_x_dum <- ((x_dum - mu_x_dum)*mu_dum - mu_x_dum*(dum - mu_dum))/(mu_dum^2)
    psi_x2_dum <- ((x2_dum - mu_x2_dum)*mu_dum - mu_x2_dum*(dum - mu_dum))/(mu_dum^2)
    psi_x3_dum <- ((x3_dum - mu_x3_dum)*mu_dum - mu_x3_dum*(dum - mu_dum))/(mu_dum^2)
    psi_F_dum <- psi_x3_dum - 3*(psi_x2_dum*mu_x_dum + mu_x2_dum*psi_x_dum) + 6*mu_x_dum^2*psi_x_dum
    psi_G_dum <- psi_x2_dum - 2*mu_x_dum*psi_x_dum
    
    infl <- (psi_F_dum*G_dum^(3/2) - 3/2*F_dum*G_dum^(1/2)*psi_G_dum)/(G_dum^3)
    infl <- as.matrix(infl, ncol = 1) 
    return(infl)
}
